domain/src/main/java/org/oppia/android/domain/exploration/ExplorationProgressController.kt

```diff
@@ -352,6 +440,112 @@ class ExplorationProgressController @Inject constructor(
     }
   }
 
+  /**
+   *  Checks if checkpointing is enabled, if checkpointing is enabled this function creates a
+   *  checkpoint with the latest progress and saves it using [ExplorationCheckpointController].
+   *
+   *  This function also waits for the save operation to complete inside of CoroutineScope and then
+   *  process the result of the save operation using the the function [processSaveCheckpointResult].
+   */
+  private fun saveExplorationCheckpoint() {
+    // do not save checkpoints if checkpointing is not enabled. This is expected to happen when
+    // the current exploration has been already completed previously.
+    if (!explorationProgress.isLightweightCheckpointingEnabled)
+      return
+    val profileId: ProfileId = explorationProgress.currentProfileId
+    val topicId: String = explorationProgress.currentTopicId
+    val storyId: String = explorationProgress.currentStoryId
+    val explorationId: String = explorationProgress.currentExplorationId
+
+    val checkpoint: ExplorationCheckpoint =
+      explorationProgress.stateDeck.createExplorationCheckpoint(
+        explorationProgress.currentExploration.version,
+        explorationProgress.currentExploration.title,
+        oppiaClock.getCurrentTimeMs()
+      )
+
+    CoroutineScope(backgroundDispatcher).launch {
+      val result = explorationCheckpointController.recordExplorationCheckpoint(
+        profileId,
+        explorationId,
+        checkpoint
+      ).retrieveData()
```

Comment:
This seems hacky. We shouldn't call retrieveData() unless we actually want to observe the result (& we probably should here--consider that any failures in this block of code will likely silently fail and not be surfaced to the application or user which will result in broken behavior that we can't investigate). In the very least, we should always log the result of a background operation.

I don't have anything against this idea in principle (that we want to entirely save the checkpoint in the background), but if it's important to observe the result then we should probably combine it with one of the existing DataProviders already being observed by StateFragment. Further, the operation to call recordExplorationCheckpoint itself isn't particularly expensive--it's already deferring execution to a background thread. To simplify, consider:
- Calling recordExplorationCheckpoint directly rather than deferring to the background dispatcher
- Update recordExplorationCheckpoint to return a Deferred (we should also make sure that this method cannot be called by the UI layer which might require some refactoring of your controller, maybe into multiple classes)
- Observe the deferred here; when it finishes you can pull the save state & notify that the ephemeral state DataProvider has an update
- Update EphemeralState to include details on the current checkpoint save status
- In the async method for the ephemeral state DataProvider, have it compute the EphemeralState such that it includes the current save state
- Use the new value in StateFragment to make decisions based on when the user wants to exit the exploration

I think the results in a net simplification:
- No need for the extra threading here
- We can avoid a call to retrieveData() by just using Deferred directly since all of this is happening in the domain layer
- We can avoid the extra DataProvider to fetch the checkpoint save status
- We can avoid blocking the UI on the above DataProvider to determine what to show the user when they try to exit

Reply:
@BenHenning even if we return a deferred from `recordExplorationCheckpoint` function we still have to use a suspend function to wait for it to complete. So in turn we will have to use a background dispatcher.

Reply:
saving works as expected using `invokeOnCompletion` on the deferred.

Reply:
For posterity: we discussed this over chat & suggested that we should use invokeOnCompletion with getCompleted rather than a DataProvider or suspend function.